Cython/Compiler/Dataclass.py

```diff
@@ -0,0 +1,685 @@
+      function is copied from the CPython implementation. (`None` isn't
+      suitable because it could also be a value for the user to pass.)
+      There's no real reason why it needs importing from the dataclasses module
+      though - it could equally be a value generated by Cython when the module loads.
+    * seen_default and the associated error message are copied directly from Python
+    * Call to user-defined __post_init__ function (if it exists) is copied from
+      CPython.
+    """
+    if not init or node.scope.lookup_here("__init__"):
+        return "", {}, []
+    # selfname behaviour copied from the cpython module
+    selfname = "__dataclass_self__" if "self" in fields else "self"
+    args = [selfname]
+
+    placeholders = {}
+    placeholder_count = [0]
+
+    # create a temp to get _HAS_DEFAULT_FACTORY
+    dataclass_module = make_dataclasses_module_callnode(node.pos)
+    has_default_factory = ExprNodes.AttributeNode(
+        node.pos,
+        obj=dataclass_module,
+        attribute=EncodedString("_HAS_DEFAULT_FACTORY")
+    )
+
+    def get_placeholder_name():
+        while True:
+            name = "INIT_PLACEHOLDER_%s" % placeholder_count[0]
+            if (name not in placeholders
+                    and name not in fields):
+                # make sure name isn't already used and doesn't
+                # conflict with a variable name (which is unlikely but possible)
+                break
+            placeholder_count[0] += 1
+        return name
+
+    default_factory_placeholder = get_placeholder_name()
+    placeholders[default_factory_placeholder] = has_default_factory
+
+    seen_default = False
+    for name, field in fields.items():
+        if not field.init.value:
+            continue
+        entry = node.scope.lookup(name)
+        if entry.annotation:
+            annotation = u": %s" % entry.annotation.string.value
+        else:
+            annotation = u""
+        assignment = u''
+        if field.default is not MISSING or field.default_factory is not MISSING:
+            seen_default = True
+            if field.default_factory is not MISSING:
+                ph_name = default_factory_placeholder
+            else:
+                ph_name = get_placeholder_name()
+                placeholders[ph_name] = field.default  # should be a node
+            assignment = u" = %s" % ph_name
+        elif seen_default:
+            error(entry.pos, ("non-default argument %s follows default argument "
+                              "in dataclass __init__") % name)
+            return "", {}, []
+
+        args.append(u"%s%s%s" % (name, annotation, assignment))
+    args = u", ".join(args)
+    func_call = u"def __init__(%s):" % args
+
+    code_lines = [
+        func_call,
+        "    pass",  # just in-case it's an empty body
+    ]
+    for name, field in fields.items():
+        if field.is_initvar:
+            continue
+        if field.default_factory is MISSING:
```

Comment:
```suggestion
        elif field.default_factory is MISSING:
```
I think making this an `elif` makes it clear to anyone reading this code that either happens.